### ISSUE TITLE
publishall needs overhaul

### ISSUE THREAD

**Issue by sunnysideup** [silverstripe/silverstripe-cms#2406]
see: https://github.com/silverstripe/silverstripe-cms/blob/4/code/Controllers/CMSMain.php#L2147-L2210

There are a few issues:

1. limit should be $pages->limit(30, $start);
(the current setup probably works, but it is not confusing as it is the exact opposite in ORM

2. it chokes when you run it due to excessive use of memory AFAIK

3. you can not run it from the command line

**Comment by sunnysideup** [comment]
Here is what I have written for our project:

```
<?php

namespace BSA\App\Tasks;

use SilverStripe\Control\Director;
use SilverStripe\Control\Controller;
use SilverStripe\Core\Environment;
use SilverStripe\Dev\BuildTask;
use SilverStripe\Security\SecurityToken;
use SilverStripe\CMS\Model\SiteTree;
use SilverStripe\Forms\FieldList;

class PublishAll extends BuildTask
{

    protected $title = 'Publish All Pages';

    protected $description = 'Publish All Pages on the Entire Site (copy content from draft to live)';

    protected $step = 10;

    public function run($request)
    {
        Environment::increaseTimeLimitTo();
        Environment::increaseMemoryLimitTo();
        if ($request->requestVar('confirm') || Director::is_cli()) {
            // Protect against CSRF on destructive action
            if (Director::is_cli() || SecurityToken::inst()->checkRequest($request)) {
                $start = 0;
                $pages = SiteTree::get()->limit($this->step, $start);
                $this->flushNow('<ol>');
                while ($pages && $pages->count()) {
                    foreach ($pages as $page) {
                        $this->flushNow('<li>publishing: '.$page->Title.'</li>');
                        $page->publishRecursive();
                        $page->destroy();
                        unset($page);
                    }
                    $start += 30;
                    $pages = SiteTree::get()->limit($this->step, $start);
                }
                $this->flushNow('</ol>');
                $this->flushNow(
                    _t(__CLASS__ . '.PUBPAGES', 'Done: Published {count} pages', array('count' => $count))
                );
            } else {
                return Controller::curr()->httpError(400);
            }
        } else {
            $response = '';
            $token = SecurityToken::inst();
            $fields = FieldList::create();
            $token->updateFieldSet($fields);
            $tokenField = $fields->first();
            $tokenHtml = ($tokenField) ? $tokenField->FieldHolder() : '';
            $publishAllDescription = _t(
                __CLASS__ . '.PUBALLFUN2',
                'Pressing this button will do the equivalent of going to every page and pressing "publish".  '
                . 'It\'s intended to be used after there have been massive edits of the content, such as when '
                . 'the site was first built.'
            );
            $response .= 
                '<h1>' . _t(__CLASS__ . '.PUBALLFUN', 'Publish All functionality') . '</h1>
                <p>' . $publishAllDescription . '</p>
                <form method="get" action="">
                    <input type="submit" name="confirm" value="'
                    . _t(__CLASS__ . '.PUBALLCONFIRM', 'Please publish every page in the site', 'Confirmation button') .'" />'
                    . $tokenHtml .
                '</form>';
        }

        $this->flushNow($response);

    }

    public function flushNow($content)
    {
        echo '';
        echo $content;
        // check that buffer is actually set before flushing
        if (ob_get_length()) {
            @ob_flush();
            @flush();
            @ob_end_flush();
        }
        @ob_start();
    }
}


```

**Comment by chillu** [comment]
I've documented this limitation on https://github.com/silverstripe/silverstripe-framework/pull/8914 and https://github.com/silverstripe/silverstripe-cms/pull/2407. In the end, this isn't a core feature we should provide, it's outlives it's usefulness - so I've marked it as deprecated. This could be replaced with a more fully featured task, ideally one that runs as a queuedjob. But I think that's scope creep for core, and can easily be solved by a community maintained module
